[SPARK-20189][DStream] Fix spark kinesis testcases to remove deprecated createStream and use Builders

yssharma opened the pull request:
## What changes were proposed in this pull request?

The spark-kinesis testcases use the KinesisUtils.createStream which are deprecated now. Modify the testcases to use the recommended KinesisInputDStream.builder instead.
This change will also enable the testcases to automatically use the session tokens automatically.

## How was this patch tested?

All the existing testcases work fine as expected with the changes.

https://issues.apache.org/jira/browse/SPARK-20189

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on external/kinesis-asl/src/test/java/org/apache/spark/streaming/kinesis/JavaKinesisStreamSuite.java:
```diff
@@ -50,48 +56,86 @@ public void testAwsCreds() {
   }
 
-  private static Function<Record, String> handler = new Function<Record, String>() {
+  public static Function1<Record, String> handler = new AbstractFunction1<Record, String>() {
```
This is now testing a Scala API in a Java test?

yssharma replied:
I agree. I however couldn't pass the org.apache.spark.api.java.function.Function to the buildWithHandler. Tried lot of other ways but only this seemed to make the api happy.

yssharma replied:
I will add more testcases for the deprecated API function createStream

yssharma replied:
The old style testcases do not work with the aws session tokens. I could add a ENV variable based check if that seems like a better solution ?

srowen replied:
I don't think this is a change you should make then. If you're switching to Scala in a Java test to avoid a deprecated API then the test isn't quite testing what it is meant to. It also sounds like it could potentially leave the existing deprecated method untested.

Are you saying there's not an un-deprecated way to call this in the Java API? That sounds like a closely related issue that should be addressed first if so.

srowen reviewed (commented):
My general comment is that we don't want to leave even deprecated methods untested.

yssharma commented:
@srowen - Have made following improvements to the patch to incorporate your points:
- Did not remove old testcases for keeping the checks on old api
- Added new testcases that test the new kinesis api
- Added an option to tag the old api testcases with flag (which can also be useful for identifying them in future easily)
- Added option to set env variable to disable the old api kinesis tests - this will help devs to run the complete testsuites on aws-session-token based infrastructure without any changes

Let me know your thoughts. Thanks.

srowen commented:
So, is the point just removing calls to a deprecated API? It doesn't seem like it needs to be this complicated. 

If a test is not specifically intended to test the deprecated API, then it should just change to use the newer undeprecated API. If that doesn't really exist, don't change that instance. Later an undeprecated alternative ought to be added, really, for callers to use as well as tests.

yssharma commented:
@srowen - Yes, 2 objectives for the patch:
- add new testcases for new api (that also tests aws session tokens)
- ability to disable old api test cases

For now I have added new test cases for the new api, so that we have the coverage for session tokens.
I could remove the part of disabling old test cases, but then while running the whole test suites there would be few failures for environments with session tokens and assumed roles, and devs might think why do we have broken test cases.

I could remove the flag option, and just keep the new test cases if that suits better. Thoughts ?

Thanks

srowen commented:
Why would you disable old API tests here? the primary purpose is not calling deprecated APIs in tests where it's not needed. Best thing is to just do that. If you happen to add a new test or two along the way to shore up coverage that's good too. I wouldn't make it more complex than that here.

yashs360 committed:
Kinesis Testcases
- add new kinesis api testcases
- add flag to disable old kinesis api testcases

SparkQA commented:
**[Test build #3635 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3635/testReport)** for PR 17506 at commit [`7f86f39`](https://github.com/apache/spark/commit/7f86f3923b4b63778f9a9b4a997367368623c7a2).

SparkQA commented:
**[Test build #3635 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3635/testReport)** for PR 17506 at commit [`7f86f39`](https://github.com/apache/spark/commit/7f86f3923b4b63778f9a9b4a997367368623c7a2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

yssharma commented:
@srowen - Thanks for the feedback. Appreciate it. Added a minimal simplified patch which fixes the testcases that fail with the old api.

yssharma commented:
The Scala style check fail because of the double spaced lines probably. But that's how the existing code was so thought of keeping it that way.

srowen commented:
@yssharma you can't check in code that fails style checks. The existing code passes the checks.

yssharma commented:
@srowen : It failed on the earlier patch with KinesisTestUtils.scala changes. This version is clean. Will wait for the next automated build :)

Build logs: https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3635/console
Commit : https://github.com/apache/spark/commit/7f86f3923b4b63778f9a9b4a997367368623c7a2

```
========================================================================
Running Scala style checks
========================================================================
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala:220: File line length exceeds 100 characters
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala:242: File line length exceeds 100 characters
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala:246: File line length exceeds 100 characters
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala:256: File line length exceeds 100 characters
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala:219:0: Use Javadoc style indentation for multiline comments
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala:227:0: Use Javadoc style indentation for multiline comments
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala:234:0: Use Javadoc style indentation for multiline comments
[error] (streaming-kinesis-asl/compile:scalastyle) errors exist
[error] Total time: 20 s, completed Apr 3, 2017 5:14:05 AM
```

yssharma commented:
@srowen - does the Jenkins re-test trigger automatically?
else, could I request a retest on this patch please ?

srowen commented:
Jenkins add to whitelist

SparkQA commented:
**[Test build #75566 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75566/testReport)** for PR 17506 at commit [`b477090`](https://github.com/apache/spark/commit/b4770900997fafbcd5e65adac11c8150211eca73).

srowen commented:
It does retrigger automatically but you need to push commits.
Hm, it may be that somehow the kinesis code isn't passing style checks and wasn't caught? because it's not generally tested unless it's changed. In any event yes definitely fix up any errors old or new.

SparkQA commented:
**[Test build #75566 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75566/testReport)** for PR 17506 at commit [`b477090`](https://github.com/apache/spark/commit/b4770900997fafbcd5e65adac11c8150211eca73).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75566/
Test PASSed.

yssharma commented:
Thanks @srowen

yssharma commented:
Is there anything else that can be done on this patch. The patch fixes all the deprecated api testcases that try to use the aws secret/id credentials instead of the builder.

yssharma commented:
@srowen do you feel this patch could be merged now ?
Thanks

srowen commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisStreamSuite.scala:
```diff
@@ -233,11 +241,15 @@ abstract class KinesisStreamTests(aggregateTestData: Boolean) extends KinesisFun
-        localTestUtils.endpointUrl, localTestUtils.regionName, InitialPositionInStream.LATEST,
-        Seconds(10), StorageLevel.MEMORY_ONLY,
-        awsCredentials.getAWSAccessKeyId, awsCredentials.getAWSSecretKey)
+      val stream = KinesisInputDStream.builder.streamingContext(ssc)
```
Just for my understanding, it's no longer necessary to pass in the credentials explicitly?

yssharma replied:
Yes thats true. The KinesisInputDStream.builder uses the Dafault credentials if no creds are passed, and the default creds work with both AWS Key and Session tokens.
So now "mvn test" works on both permanent aws keys & session based token environments.

srowen commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -267,7 +267,7 @@ object KinesisInputDStream {
         checkpointInterval.getOrElse(ssc.graph.batchDuration),
         storageLevel.getOrElse(DEFAULT_STORAGE_LEVEL),
-        handler,
+        ssc.sc.clean(handler),
```
Is this related?

yssharma replied:
Yes this is required.
KinesisUtils used to send a cleaned Handler while creating stream, so we do this in the KinesisInputDstream now.

`val cleanedHandler = ssc.sc.clean(messageHandler)`

srowen reviewed (commented):
I don't know enough to really evaluate

srowen commented:
Merged to master

yssharma commented:
Thanks @srowen
